### DISCUSSION TITLE
build failed with error unexpected end of JSON input

### DISCUSSION THREAD

**Discussion by ttys000** [rbenv/ruby-build#2398] [Build failures]
Hi guys,

I want to install ruby on my Mac via a version manager like rbenv or asdf, but it's all failed, they both use ruby-build under the hood.

Check the build log I found the following error message:

```sh
2024/06/01 01:16:10 unmarshal message: unexpected end of JSON input
make: *** [main.o] Error 1
make: *** Waiting for unfinished jobs....
Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-darwin23.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
external command failed with status 2
```

Does anyone know about this error? Or how can I get more detailed information to fix it.

## Environment

- chip: Apple M1
- macOS version: Sonoma 14.5 (23F79)
- xcode-select -p: /Applications/Xcode.app/Contents/Developer
- clang version: Apple clang version 15.0.0 (clang-1500.3.9.4)
- rustc version: 1.78.0

## Terminal Output

```sh
asdf install ruby 3.3.1 -v

==> Downloading ruby-3.3.1.tar.gz...
-> curl -q -fL -o ruby-3.3.1.tar.gz https://cache.ruby-lang.org/pub/ruby/3.3/ruby-3.3.1.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 21.0M  100 21.0M    0     0  17.7M      0  0:00:01  0:00:01 --:--:-- 17.7M
==> Installing ruby-3.3.1...
ruby-build: using libyaml from homebrew
ruby-build: using gmp from homebrew
-> ./configure "--prefix=$HOME/.asdf/installs/ruby/3.3.1" --enable-shared --with-libyaml-dir=/opt/homebrew/opt/libyaml --with-gmp-dir=/opt/homebrew/opt/gmp --with-ext=openssl,psych,+ --with-readline-dir=/opt/homebrew/opt/readline --with-openssl-dir=/opt/homebrew/opt/openssl@3
-> make -j 8

BUILD FAILED (macOS 14.5 on arm64 using ruby-build 20240517)

You can inspect the build directory at /var/folders/d5/89cq_k3j2g73crqs3rt_gnfc0000gn/T/ruby-build.20240601011521.44208.J3Iire
See the full build log at /var/folders/d5/89cq_k3j2g73crqs3rt_gnfc0000gn/T/ruby-build.20240601011521.44208.log
```
## Build Log

```sh
cd /var/folders/d5/89cq_k3j2g73crqs3rt_gnfc0000gn/T/ruby-build.20240601011521.44208.J3Iire
==> Downloading ruby-3.3.1.tar.gz...
-> curl -q -fL -o ruby-3.3.1.tar.gz https://cache.ruby-lang.org/pub/ruby/3.3/ruby-3.3.1.tar.gz
-> tar xzf ruby-3.3.1.tar.gz
cd /var/folders/d5/89cq_k3j2g73crqs3rt_gnfc0000gn/T/ruby-build.20240601011521.44208.J3Iire/ruby-3.3.1
==> Installing ruby-3.3.1...
-> ./configure "--prefix=$HOME/.asdf/installs/ruby/3.3.1" --enable-shared --with-libyaml-dir=/opt/homebrew/opt/libyaml --with-gmp-dir=/opt/homebrew/opt/gmp --with-ext=openssl,psych,+ --with-readline-dir=/opt/homebrew/opt/readline --with-openssl-dir=/opt/homebrew/opt/openssl@3
checking for ruby... /usr/bin/ruby
tool/config.guess already exists
tool/config.sub already exists
checking build system type... aarch64-apple-darwin23.5.0
checking host system type... aarch64-apple-darwin23.5.0
checking target system type... aarch64-apple-darwin23.5.0
checking for ld... ld
checking for llvm-ar... no
checking for clang++... clang++
checking for llvm-nm... no
checking for llvm-objcopy... no
checking for llvm-objdump... no
checking for llvm-ranlib... no
checking for llvm-strip... no
checking for gcc... clang
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether clang accepts -g... yes
checking for clang option to enable C11 features... none needed
checking whether the compiler supports GNU C++... yes
checking whether clang++ accepts -g... yes
checking for clang++ option to enable C++11 features... none needed
checking how to run the C preprocessor... clang -E
checking for ranlib... ranlib
checking for gar... (cached) /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ar
checking for gas... no
checking for as... as
checking for gnm... (cached) /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/nm
checking for gobjcopy... (cached) :
checking for gobjdump... no
checking for objdump... objdump
checking for gstrip... no
checking for strip... strip
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for wchar.h... yes
checking for minix/config.h... no
checking for vfork.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking whether _XOPEN_SOURCE should be defined... no
checking if minimum required OS X version is supported... yes
checking whether the linker is GNU ld... no
checking whether clang -E accepts -o... yes
checking for real target cpu... arm64
checking arch option... -arch arm64
checking for /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ar flags... rcu
checking whether ln -s works... yes
checking whether make sets $(MAKE)... no
checking for a BSD-compatible install... /opt/homebrew/bin/ginstall -c
checking for a race-free mkdir -p... /opt/homebrew/bin/gmkdir -p
checking for dtrace... dtrace
checking for dot... dot
checking for doxygen... no
checking for pkg-config... pkg-config
checking whether it is Android... no
checking for cd using physical directory... cd -P
checking whether CFLAGS is valid... yes
checking whether LDFLAGS is valid... yes
checking whether ELF binaries are produced... no
checking whether -Wl,-rpath,. is accepted as LDFLAGS... yes
checking whether -fdeclspec is accepted as LDFLAGS... yes
checking whether -fdeclspec is accepted as CXXFLAGS... yes
checking whether -Wextra-tokens is accepted as CFLAGS... yes
checking whether -Wdeprecated-declarations is accepted as CFLAGS... yes
checking whether -Wdivision-by-zero is accepted as CFLAGS... yes
checking whether -Wdiv-by-zero is accepted as CFLAGS... yes
checking whether -Wduplicated-cond is accepted as CFLAGS... no
checking whether -Wimplicit-function-declaration is accepted as CFLAGS... yes
checking whether -Wimplicit-int is accepted as CFLAGS... yes
checking whether -Wpointer-arith is accepted as CFLAGS... yes
checking whether -Wshorten-64-to-32 is accepted as CFLAGS... yes
checking whether -Wwrite-strings is accepted as CFLAGS... yes
checking whether -Wold-style-definition is accepted as CFLAGS... yes
checking whether -Wimplicit-fallthrough=0 is accepted as CFLAGS... no
checking whether -Wmissing-noreturn is accepted as CFLAGS... yes
checking whether -Wno-cast-function-type is accepted as CFLAGS... yes
checking whether -Wno-constant-logical-operand is accepted as CFLAGS... yes
checking whether -Wno-long-long is accepted as CFLAGS... yes
checking whether -Wno-missing-field-initializers is accepted as CFLAGS... yes
checking whether -Wno-overlength-strings is accepted as CFLAGS... yes
checking whether -Wno-packed-bitfield-compat is accepted as CFLAGS... no
checking whether -Wno-parentheses-equality is accepted as CFLAGS... yes
checking whether -Wno-self-assign is accepted as CFLAGS... yes
checking whether -Wno-tautological-compare is accepted as CFLAGS... yes
checking whether -Wno-unused-parameter is accepted as CFLAGS... yes
checking whether -Wno-unused-value is accepted as CFLAGS... yes
checking whether -Wsuggest-attribute=format is accepted as CFLAGS... no
checking whether -Wsuggest-attribute=noreturn is accepted as CFLAGS... no
checking whether -Wunused-variable is accepted as CFLAGS... yes
checking whether -diag-disable=175,188,1684,2259,2312 is accepted as CFLAGS... no
checking whether -Wno-maybe-uninitialized is accepted as CFLAGS... no
checking whether -Wmisleading-indentation is accepted as CFLAGS... yes
checking whether -Wundef is accepted as CFLAGS... yes
checking whether -Wall -Wextra is accepted as CFLAGS... yes
checking whether -Qunused-arguments is accepted as CFLAGS... yes
checking whether -O3 -D_FORTIFY_SOURCE=2 is accepted as CFLAGS... yes
checking whether -fstack-protector-strong is accepted as CFLAGS... yes
checking whether -fstack-protector-strong is accepted as LDFLAGS... yes
checking for -fstack-protector... -fstack-protector-strong
checking whether -Wl,--compress-debug-sections=zlib is accepted as LDFLAGS... no
checking whether -fno-strict-overflow is accepted as CFLAGS... yes
checking whether -ggdb3 is accepted as CFLAGS... yes
checking whether -fvisibility=hidden is accepted as CFLAGS... yes
checking whether -fno-fast-math is accepted as CFLAGS... yes
checking whether -fexcess-precision=standard is accepted as CFLAGS... no
checking whether -fp-model precise is accepted as CFLAGS... no
checking whether compiler has statement and declarations in expressions... yes
checking for pthread.h... yes
checking for pthread_create in -lthr... no
checking for pthread_create in -lpthread... yes
checking for pthread_np.h... no
checking thread model... pthread
checking whether syscall(2) is deprecated... yes
checking for broken crypt with 8bit chars... yes
checking for codesign... codesign
checking for dsymutil... dsymutil
checking for crt_externs.h... yes
checking whether CXXFLAGS is valid... yes
checking for crypt in -lcrypt... (cached) no
checking for dlopen in -ldl... yes
checking for shutdown in -lsocket... no
checking for dirent.h that defines DIR... yes
checking for library containing opendir... none required
checking for _Bool... yes
checking for stdbool.h that conforms to C99... yes
checking for sys/wait.h that is POSIX.1 compatible... yes
checking for afunix.h... no
checking for atomic.h... no
checking for copyfile.h... yes
checking for direct.h... no
checking for grp.h... yes
checking for fcntl.h... yes
checking for float.h... yes
checking for ieeefp.h... no
checking for intrinsics.h... no
checking for langinfo.h... yes
checking for limits.h... yes
checking for locale.h... yes
checking for malloc.h... no
checking for malloc/malloc.h... yes
checking for malloc_np.h... no
checking for net/socket.h... no
checking for process.h... no
checking for pwd.h... yes
checking for sanitizer/asan_interface.h... yes
checking for sanitizer/msan_interface.h... yes
checking for setjmpex.h... no
checking for stdalign.h... yes
checking for stdio.h... (cached) yes
checking for sys/attr.h... yes
checking for sys/eventfd.h... no
checking for sys/fcntl.h... yes
checking for sys/file.h... yes
checking for sys/id.h... no
checking for sys/ioctl.h... yes
checking for sys/mkdev.h... no
checking for sys/param.h... yes
checking for sys/prctl.h... no
checking for sys/random.h... yes
checking for sys/resource.h... yes
checking for sys/select.h... yes
checking for sys/sendfile.h... no
checking for sys/socket.h... yes
checking for sys/syscall.h... (cached) no
checking for sys/sysmacros.h... no
checking for sys/time.h... yes
checking for sys/times.h... yes
checking for sys/uio.h... yes
checking for sys/utime.h... no
checking for syscall.h... (cached) no
checking for time.h... yes
checking for ucontext.h... yes
checking for utime.h... yes
checking for sys/epoll.h... no
checking for sys/event.h... yes
checking for gmp.h... no
checking for special C compiler options needed for large files... no
checking for _FILE_OFFSET_BITS value needed for large files... no
checking whether byte ordering is bigendian... no
checking for an ANSI C-conforming const... yes
checking whether char is unsigned... no
checking for inline... inline
checking for working volatile... yes
checking for typeof syntax and keyword spelling... typeof
checking for C/C++ restrict keyword... __restrict__
checking for long long... yes
checking for off_t... yes
checking char bit... 8
checking size of int... 4
checking size of short... 2
checking size of long... 8
checking size of long long... 8
checking size of __int64... 0
checking size of __int128... 16
checking size of off_t... 8
checking size of void*... 8
checking size of float... 4
checking size of double... 8
checking size of time_t... 8
checking size of clock_t... 8
checking word size... 8
checking packed struct attribute... x __attribute__((packed))
checking whether -Wno-address-of-packed-member is accepted as CFLAGS... yes
checking for printf prefix for long long... ll
checking for pid_t... yes
checking for convertible type of pid_t... INT
checking for uid_t... yes
checking for convertible type of uid_t... UINT
checking for gid_t... yes
checking for convertible type of gid_t... UINT
checking for time_t... yes
checking for convertible type of time_t... LONG
checking for dev_t... yes
checking for convertible type of dev_t... INT
checking for mode_t... yes
checking for convertible type of mode_t... USHORT
checking for rlim_t... yes
checking for convertible type of rlim_t... ULL
checking for off_t... (cached) yes
checking for convertible type of off_t... LL
checking for clockid_t... yes
checking for convertible type of clockid_t... UINT
checking for variable length macro... yes
checking if _Alignof() works... yes
checking for CONSTFUNC function attribute... __attribute__ ((__const__)) x
checking for PUREFUNC function attribute... __attribute__ ((__pure__)) x
checking for NORETURN function attribute... __attribute__ ((__noreturn__)) x
checking for DEPRECATED function attribute... __attribute__ ((__deprecated__)) x
checking for DEPRECATED_BY function attribute... __attribute__ ((__deprecated__("by "#n))) x
checking for NOINLINE function attribute... __attribute__ ((__noinline__)) x
checking for ALWAYS_INLINE function attribute... __attribute__ ((__always_inline__)) x
checking for NO_SANITIZE function attribute... __attribute__ ((__no_sanitize__(san))) x
checking for NO_SANITIZE_ADDRESS function attribute... __attribute__ ((__no_sanitize_address__)) x
checking for NO_ADDRESS_SAFETY_ANALYSIS function attribute... __attribute__ ((__no_address_safety_analysis__)) x
checking for WARN_UNUSED_RESULT function attribute... __attribute__ ((__warn_unused_result__)) x
checking for MAYBE_UNUSED function attribute... __attribute__ ((__unused__)) x
checking for ERRORFUNC function attribute... __attribute__ ((__error__ mesg)) x
checking for WARNINGFUNC function attribute... __attribute__ ((__warning__ mesg)) x
checking for WEAK function attribute... __attribute__ ((__weak__)) x
checking for __attribute__((__depreacted__(msg))) in C++... yes
checking for std::nullptr_t... yes
checking for FUNC_STDCALL function attribute... x
checking for FUNC_CDECL function attribute... __attribute__ ((__cdecl__)) x
checking for FUNC_FASTCALL function attribute... x
checking for FUNC_UNOPTIMIZED function attribute... x
checking for FUNC_MINIMIZED function attribute... x
checking for function alias... no
checking for __atomic builtins... yes
checking for __sync builtins... yes
checking for __builtin_unreachable... yes
checking for __assume... no
checking for exported function attribute... __attribute__ ((__visibility__("default")))
checking for RUBY_FUNC_NONNULL function attribute... __attribute__ ((__nonnull__(n))) x
checking for function name string predefined identifier... __func__
checking if enum over int is allowed... yes
checking for clang options needed to detect all undeclared functions... none needed
checking whether sys_nerr is declared... yes
checking whether getenv is declared... yes
checking for size_t... yes
checking size of size_t... 8
checking size of ptrdiff_t... 8
checking size of dev_t... 4
checking for printf prefix for size_t... z
checking for printf prefix for ptrdiff_t... t
checking for struct stat.st_blksize... yes
checking for struct stat.st_blocks... yes
checking for struct stat.st_rdev... yes
checking size of struct stat.st_size... SIZEOF_OFF_T
checking size of struct stat.st_blocks... SIZEOF_OFF_T
checking size of struct stat.st_ino... SIZEOF_LONG
checking size of struct stat.st_dev... SIZEOF_DEV_T
checking size of struct stat.st_rdev... SIZEOF_DEV_T
checking for struct stat.st_atim... no
checking for struct stat.st_atimespec... yes
checking for struct stat.st_atimensec... no
checking for struct stat.st_mtim... no
checking for struct stat.st_mtimespec... yes
checking for struct stat.st_mtimensec... no
checking for struct stat.st_ctim... no
checking for struct stat.st_ctimespec... yes
checking for struct stat.st_ctimensec... no
checking for struct stat.st_birthtimespec... yes
checking for struct timeval... yes
checking size of struct timeval.tv_sec... SIZEOF_TIME_T
checking for struct timespec... yes
checking for struct timezone... yes
checking for fd_mask... yes
checking for int8_t... yes
checking size of int8_t... 1
checking for uint8_t... yes
checking size of uint8_t... 1
checking for int16_t... yes
checking size of int16_t... 2
checking for uint16_t... yes
checking size of uint16_t... 2
checking for int32_t... yes
checking size of int32_t... 4
checking for uint32_t... yes
checking size of uint32_t... 4
checking for int64_t... yes
checking size of int64_t... 8
checking for uint64_t... yes
checking size of uint64_t... 8
checking for int128_t... __int128
checking for uint128_t... unsigned __int128
checking for intptr_t... yes
checking size of intptr_t... 8
checking for uintptr_t... yes
checking size of uintptr_t... 8
checking for printf prefix for intptr_t... l
checking for ssize_t... yes
checking size of ssize_t... 8
checking for printf prefix for int64_t... ll
checking for stack end address... no
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for uid_t in sys/types.h... (cached) yes
checking type of array argument to getgroups... (cached) gid_t
checking for working alloca.h... yes
checking for alloca... yes
checking for dynamic size alloca... ok
checking for working memcmp... yes
checking for dup... yes
checking for dup2... yes
checking for acosh... yes
checking for cbrt... yes
checking for crypt... yes
checking for erf... yes
checking for explicit_bzero... no
checking for ffs... yes
checking for flock... yes
checking for hypot... yes
checking for lgamma_r... yes
checking for memmove... yes
checking for nan... yes
checking for nextafter... yes
checking for setproctitle... no
checking for strchr... yes
checking for strerror... yes
checking for strlcat... yes
checking for strlcpy... yes
checking for strstr... yes
checking for tgamma... yes
checking for sys/pstat.h... no
checking for pid_t... (cached) yes
checking for fork... yes
checking for vfork... (cached) no
checking for working fork... yes
checking for __syscall... (cached) no
checking for _longjmp... yes
checking for arc4random_buf... yes
checking for atan2l... yes
checking for atan2f... yes
checking for chmod... yes
checking for chown... yes
checking for chroot... yes
checking for chsize... no
checking for clock_gettime... yes
checking for copy_file_range... no
checking for cosh... yes
checking for crypt_r... no
checking for daemon... (cached) no
checking for dirfd... yes
checking for dl_iterate_phdr... no
checking for dlopen... yes
checking for dladdr... yes
checking for dup3... no
checking for eaccess... no
checking for endgrent... yes
checking for eventfd... no
checking for execl... yes
checking for execle... yes
checking for execv... yes
checking for execve... yes
checking for explicit_memset... no
checking for fcopyfile... yes
checking for fchdir... yes
checking for fchmod... yes
checking for fchown... yes
checking for fcntl... yes
checking for fdatasync... (cached) no
checking for fdopendir... yes
checking for fgetattrlist... yes
checking for fmod... yes
checking for fstatat... yes
checking for fsync... yes
checking for ftruncate... yes
checking for ftruncate64... no
checking for getattrlist... yes
checking for getcwd... yes
checking for getegid... yes
checking for getentropy... yes
checking for geteuid... yes
checking for getgid... yes
checking for getgidx... no
checking for getgrnam... yes
checking for getgrnam_r... yes
checking for getgroups... yes
checking for getlogin... yes
checking for getlogin_r... yes
checking for getpgid... yes
checking for getpgrp... yes
checking for getppid... yes
checking for getpriority... yes
checking for getpwnam... yes
checking for getpwnam_r... yes
checking for getpwuid... yes
checking for getpwuid_r... yes
checking for getrandom... no
checking for getresgid... no
checking for getresuid... no
checking for getrlimit... yes
checking for getsid... yes
checking for gettimeofday... yes
checking for getuid... yes
checking for getuidx... no
checking for gmtime_r... yes
checking for grantpt... yes
checking for initgroups... yes
checking for ioctl... yes
checking for isfinite... no
checking for issetugid... yes
checking for kill... yes
checking for killpg... yes
checking for lchmod... yes
checking for lchown... yes
checking for link... yes
checking for llabs... yes
checking for lockf... yes
checking for log2... yes
checking for lstat... yes
checking for lutimes... yes
checking for malloc_usable_size... no
checking for malloc_size... yes
checking for malloc_trim... no
checking for mblen... yes
checking for memalign... no
checking for memset_s... yes
checking for writev... yes
checking for memrchr... no
checking for memmem... yes
checking for mkfifo... yes
checking for mknod... yes
checking for mktime... yes
checking for mmap... yes
checking for mremap... no
checking for openat... yes
checking for pclose... yes
checking for pipe... yes
checking for pipe2... no
checking for poll... yes
checking for popen... yes
checking for posix_fadvise... no
checking for posix_madvise... yes
checking for posix_memalign... yes
checking for ppoll... no
checking for pread... yes
checking for pwrite... yes
checking for qsort_r... yes
checking for qsort_s... no
checking for readlink... yes
checking for realpath... yes
checking for round... yes
checking for sched_getaffinity... no
checking for seekdir... yes
checking for select_large_fdset... no
checking for sendfile... yes
checking for setegid... yes
checking for setenv... yes
checking for seteuid... yes
checking for setgid... yes
checking for setgroups... yes
checking for setpgid... yes
checking for setpgrp... yes
checking for setregid... yes
checking for setresgid... no
checking for setresuid... no
checking for setreuid... yes
checking for setrgid... yes
checking for setrlimit... yes
checking for setruid... yes
checking for setsid... yes
checking for setuid... yes
checking for shutdown... yes
checking for sigaction... yes
checking for sigaltstack... yes
checking for sigprocmask... yes
checking for sinh... yes
checking for snprintf... yes
checking for spawnv... no
checking for symlink... yes
checking for syscall... (cached) no
checking for sysconf... yes
checking for system... yes
checking for tanh... yes
checking for telldir... yes
checking for timegm... yes
checking for times... yes
checking for truncate... yes
checking for truncate64... no
checking for tzset... yes
checking for umask... yes
checking for unsetenv... yes
checking for utimensat... yes
checking for utimes... yes
checking for wait4... yes
checking for waitpid... yes
checking for __cospi... yes
checking for __sinpi... yes
checking if getcwd allocates buffer if NULL is given... yes
checking for __builtin_alloca_with_align... yes
checking for __builtin_assume_aligned... yes
checking for __builtin_bswap16... yes
checking for __builtin_bswap32... yes
checking for __builtin_bswap64... yes
checking for __builtin_popcount... yes
checking for __builtin_popcountll... yes
checking for __builtin_clz... yes
checking for __builtin_clzl... yes
checking for __builtin_clzll... yes
checking for __builtin_ctz... yes
checking for __builtin_ctzll... yes
checking for __builtin_add_overflow... yes
checking for __builtin_sub_overflow... yes
checking for __builtin_mul_overflow... yes
checking for __builtin_mul_overflow_p... no
checking for __builtin_constant_p... yes
checking for __builtin_choose_expr... yes
checking for __builtin_choose_expr_constant_p... yes
checking for __builtin_types_compatible_p... yes
checking for __builtin_trap... yes
checking for __builtin_expect... yes
checking for __builtin_mul_overflow with long long arguments... yes
checking whether qsort_r is GNU version... no
checking whether qsort_r is BSD version... yes
checking whether atan2 handles Inf as C99... yes
checking for clock_getres... yes
checking for timer_create in -lrt... no
checking for timer_settime in -lrt... no
checking for unsetenv returns a value... yes
checking whether struct tm is in sys/time.h or time.h... time.h
checking for struct tm.tm_zone... yes
checking for struct tm.tm_gmtoff... yes
checking for external int daylight... yes
checking for negative time_t for gmtime(3)... yes
checking for localtime(3) overflow correctly... yes
checking for sig_t... yes
checking whether right shift preserve sign bit... yes
checking for copy_file_range... no
checking whether _SC_CLK_TCK is supported... yes
checking stack growing direction on arm64... -1
checking native coroutine implementation for arm64-darwin23... arm64
checking whether pthread_t is scalar type... yes
checking for sched_yield... yes
checking for pthread_attr_setinheritsched... yes
checking for pthread_attr_get_np... no
checking for pthread_attr_getstack... yes
checking for pthread_attr_getguardsize... yes
checking for pthread_get_stackaddr_np... yes
checking for pthread_get_stacksize_np... yes
checking for thr_stksegment... no
checking for pthread_stackseg_np... no
checking for pthread_getthrds_np... no
checking for pthread_condattr_setclock... no
checking for pthread_setname_np... yes
checking for pthread_set_name_np... no
checking for pthread_sigmask... yes
checking for pthread_getattr_np... no
checking arguments of pthread_setname_np... (name)
checking for thread-local storage sepcifier... _Thread_local
checking if mcontext_t is a pointer... yes
checking for getcontext... (cached) no
checking for setcontext... (cached) no
checking if fork works with pthread... yes
checking for sys/user.h... yes
checking whether PAGE_SIZE is compile-time const... no
checking ioctl request type... unsigned long
checking for mach-o/loader.h... yes
checking whether OS depend dynamic link works... yes
checking whether -Wl,-multiply_defined,suppress is accepted as LDFLAGS... no
checking whether -Wl,-undefined,dynamic_lookup is accepted as LDFLAGS... yes
checking whether -Wl,-undefined,dynamic_lookup is accepted for bundle... yes
checking for libproc.h... yes
checking for execinfo.h... yes
checking for backtrace in -lexecinfo... no
checking for libunwind.h... yes
checking for mach/task.h... yes
checking for mach/mach_init.h... yes
checking for mach/mach_port.h... yes
checking for backtrace... yes
checking for broken backtrace... yes
checking for valgrind/memcheck.h... no
checking for strip flags... -A -n
checking whether -Wl,--no-as-needed is accepted as LDFLAGS... no
checking whether -Wl,--no-undefined is accepted as LDFLAGS... no
checking whether wrapper for DYLD_LIBRARY_PATH is needed... yes
checking whether dtrace USDT is available... yes(-xnolibs)
checking whether dtrace needs post processing... no
checking for __builtin_setjmp... (cached) no
checking for _setjmpex as a macro or function... no
checking for _setjmp as a macro or function... yes
checking for sigsetjmp as a macro or function... yes
checking for setjmp type... sigsetjmp,0
checking for rustc... rustc
checking whether rustc works for YJIT... yes
checking for prefix of external symbols... _
checking for default symbols in empty shared library... 
checking whether Security framework is needed... no
checking if make is GNU make... no
checking for safe null command for make... :
.ext/include/arm64-darwin23/ruby/config.h updated
configure: ruby library version = 3.3.0
configure: creating ./config.status
config.status: creating Makefile
---
Configuration summary for ruby version 3.3.1

   * Installation prefix: /Users/peng/.asdf/installs/ruby/3.3.1
   * exec prefix:         ${prefix}
   * arch:                arm64-darwin23
   * site arch:           ${arch}
   * RUBY_BASE_NAME:      ruby
   * enable shared:       yes
   * ruby lib prefix:     ${libdir}/${RUBY_BASE_NAME}
   * site libraries path: ${rubylibprefix}/${sitearch}
   * vendor path:         ${rubylibprefix}/vendor_ruby
   * target OS:           darwin23
   * compiler:            clang
   * with thread:         pthread
   * with coroutine:      arm64
   * enable shared libs:  yes
   * dynamic library ext: bundle
   * CFLAGS:              -fdeclspec ${optflags} ${debugflags} ${warnflags}
   * LDFLAGS:             -L. -fstack-protector-strong \
                          -L/opt/homebrew/opt/gmp/lib
   * DLDFLAGS:            -L/opt/homebrew/opt/gmp/lib \
                          -Wl,-undefined,dynamic_lookup
   * optflags:            -O3 -fno-fast-math
   * debugflags:          -ggdb3
   * warnflags:           -Wall -Wextra -Wextra-tokens \
                          -Wdeprecated-declarations -Wdivision-by-zero \
                          -Wdiv-by-zero -Wimplicit-function-declaration \
                          -Wimplicit-int -Wpointer-arith -Wshorten-64-to-32 \
                          -Wwrite-strings -Wold-style-definition \
                          -Wmissing-noreturn -Wno-cast-function-type \
                          -Wno-constant-logical-operand -Wno-long-long \
                          -Wno-missing-field-initializers \
                          -Wno-overlength-strings -Wno-parentheses-equality \
                          -Wno-self-assign -Wno-tautological-compare \
                          -Wno-unused-parameter -Wno-unused-value \
                          -Wunused-variable -Wmisleading-indentation -Wundef
   * strip command:       strip -A -n
   * install doc:         rdoc
   * YJIT support:        yes
   * RJIT support:        yes
   * man page type:       doc
   * BASERUBY -v:         ruby 2.6.10p210 (2022-04-12 revision 67958) \
                          [universal.arm64e-darwin23]

---
-> make -j 8
Makefile:21362: warning: overriding commands for target `ruby'
Makefile:300: warning: ignoring old commands for target `ruby'
	BASERUBY = /usr/bin/ruby --disable=gems
	CC = clang
	LD = ld
	LDSHARED = clang -dynamiclib
	CFLAGS = -fdeclspec -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wextra-tokens -Wdeprecated-declarations -Wdivision-by-zero -Wdiv-by-zero -Wimplicit-function-declaration -Wimplicit-int -Wpointer-arith -Wshorten-64-to-32 -Wwrite-strings -Wold-style-definition -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wunused-variable -Wmisleading-indentation -Wundef  -fno-common -pipe -arch arm64
	XCFLAGS = -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -fstack-protector-strong -fno-strict-overflow -fvisibility=hidden -DRUBY_EXPORT -I. -I.ext/include/arm64-darwin23 -I./include -I. -I./prism -I./enc/unicode/15.0.0 -I/opt/homebrew/opt/gmp/include 
	CPPFLAGS = -D_XOPEN_SOURCE -D_DARWIN_C_SOURCE -D_DARWIN_UNLIMITED_SELECT -D_REENTRANT   
	DLDFLAGS = -L/opt/homebrew/opt/gmp/lib -Wl,-undefined,dynamic_lookup -install_name /Users/peng/.asdf/installs/ruby/3.3.1/lib/libruby.3.3.dylib -compatibility_version 3.3 -current_version 3.3.1  -fstack-protector-strong -framework CoreFoundation  -fstack-protector-strong -framework CoreFoundation  -arch arm64
	SOLIBS = -ldl -lobjc -lpthread 
	LANG = en_US.UTF-8
	LC_ALL = 
	LC_CTYPE = en_US.UTF-8
	MFLAGS = - --jobserver-fds=5,6 -j
	RUSTC = rustc
	YJIT_RUSTC_ARGS = --crate-name=yjit --crate-type=staticlib --edition=2021 -g -C lto=thin -C opt-level=3 -C overflow-checks=on '--out-dir=/var/folders/d5/89cq_k3j2g73crqs3rt_gnfc0000gn/T/ruby-build.20240601011521.44208.J3Iire/ruby-3.3.1/yjit/target/release/' ./yjit/src/lib.rs
2024/06/01 01:16:10 unmarshal message: unexpected end of JSON input
make: *** [main.o] Error 1
make: *** Waiting for unfinished jobs....
Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-darwin23.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
external command failed with status 2

```

**Comment by ameykusurkar** [comment]
Hi @ttys000 I've been having the same issue as you, but I managed to fix it!

**TL;DR** the issue was _completely unrelated_ to `ruby-build`, but rather `echo` (which is run by `make`) was waiting on JSON input for some reason, and turns out this was because I had a custom `echo` binary. I removed the binary and let it default to `/bin/echo`, which solved the issue.

### Debugging the issue

Similar to you, I was also seeing this:
```shell
2024/06/01 01:16:10 unmarshal message: unexpected end of JSON input
make: *** [main.o] Error 1
make: *** Waiting for unfinished jobs....
Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-darwin23.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
external command failed with status 2
```

The thing that was really suspicious to me was:
```
2024/06/01 01:16:10 unmarshal message: unexpected end of JSON input
```
which is an error message that you see when parsing JSON using [Golang's `Unmarshall`](https://pkg.go.dev/encoding/json#Unmarshal). I was pretty sure that nothing in this toolchain actually uses Go. I narrowed down the issue to `make` and I managed to reproduce the issue with a simpler `Makefile` that just did an `echo hello`, which gave me the same error (although I had to type something to stdin first to trigger the error).

However, `env -i PATH=/usr/bin:/bin make` was working fine, which means that I had a weird version of `echo` in my `PATH` (there were no other commands in my test). In fact, `which echo` showed me this:
```
/Users/ameykusurkar/.asdf/installs/golang/1.22.3/packages/bin/echo
```
which is really weird! Not the `echo` I wanted to use.

I then ran:
```
rm /Users/ameykusurkar/.asdf/installs/golang/1.22.3/packages/bin/echo
asdf reshim golang
```

Which fixed the issue! It might be that you have the exact same issue (ie. a bad `echo`), but if not I'd check your `PATH` to see if you have a Go binary running instead of some standard binary that is used in this `ruby-build` install process.

**Reply by mislav** [reply]
@ameykusurkar: Thanks for sharing your own experience and solution!

@ttys000: I also think that this is caused by something external to ruby-build and Ruby's own build system, like some binary in your PATH that shadows a system binary of the same name. Let us know if you find it! 👍 

Tip: you can set the `RUBY_MAKE_OPTS=V=1` environment variable while invoking ruby-build (or `adsf install`) to have Ruby's build system print external commands as it invokes them. (This will make the build log _very_ verbose.)
